doc/bmp/bmp.md

```diff
@@ -0,0 +1,373 @@
+| **Term**                 | **Meaning**                         | **Link**                         |
+|--------------------------|-------------------------------------|--------------------------|
+| BMP                      | BGP Monitor Protocol                |https://datatracker.ietf.org/doc/html/rfc7854 |
+| OpenBMP                  | Open Source BGP Monitoring Protocol (OpenBMP) Collection Framework    |https://github.com/SNAS/openbmp|
+| BGP/FRR                  | SONiC BGP container and bgpd daemon which will collborate with BMP here   |https://github.com/sonic-net/sonic-frr|
+| GNMI                     | gRPC Network Management Interface   |https://github.com/openconfig/gnmi|
+
+
+# 1 Requirements Overview
+
+## 1.1 Functional requirements
+
+Now SONiC supports standard BGP session established with BGP neighbors, but there's only limited BGP info readable in database, which makes some cases difficult to debug, we need to dump more BGP data, meanwhile we're also expecting some live debug option which could dump BGP related info on demand and improve the debuggability.
+
+At a high level the following data should be supported:
+
+- BGP neighbor table, especially capability like MPBGP support, graceful_restart support, etc
+- BGP in and out RIB table, now only ifname and nexthop are available in ROUTE_TABLE, more data like as_path/asn/origin... and in/out rib info need to be added.
+- Number of prefixes received and sent.
+
+Moreover, now if we want to monitor BGP session data of SONiC, now we need to create standard BGP session with SONiC device, the restriction is that the data monitored is not identical with the data which SONiC converses with neighbors. With BMP data being available on SONiC, we should be able to monitor more BGP live data. 
+
+As a result of this implementation, BGP listeners should be able to use StreamingTelemetry "on change" subscription channel, which is state based monitoring and more efficient than operation based monitoring as BGP standard session.
+
+
+## 1.2 CLI requirements
+- Use Config cli to enable bmp table population as granularity of table
+- Use show to query specific bmp table data
+
+## 1.3 Scalability and Default Values
+
+As a phase #1 scalability requirement, the proposed support is to have bgp neighbor table enabled by default, considering the device set like 7050qx (4G RAM), but it should be based on final evaluation, more details will be discussed in [4 Resource usage and Test plan](#4-resource-usage-and-test-plan)
+
+
+## 1.4 Warm Restart requirements
+No special handling for Warm restart support.
+
+# 2 Architecture Design
+
+The following are the architecture changes. 
+
+## 2.1 High-Level Architecture
+As above context mentioned, we need to bring up new container as bmp on SONiC, after change the relevant component architecture is as below:
+<img src="images/architecture_diagram.png" alt="Architecture Diagram" width="800">  
+
+- Add new bmp container, which has limited resource control. 
```

Comment:
Do we new docker for running the "openBMPd" daemon? 

I understand that the plan is to have, 

1) config "feature" command to enable/disable this.
2) Limit CPU resource usage of the new docker . 

Even if we leverage the existing frr container to run openBMPd, there are alternate ways to achieve the above, for instance #1 we could have a "config bmp <enable|disable>"  which could internally start/stop the openBMPd process. And for #2 we could use command such as "nice/cpulimit" to set process usage limits. 

From what i have seen adding a new container (base docker, with no overhead app) increases RAM usage by 20 to 30 MB, so just want to ensure if we really need a new container here.

Reply:
Thanks, it's indeed feasible as either way, and some tradeoff just exists as consideration.
The advantage of using dedicated container is that we can avoid packaging bmp feature onto some slim images if DUT resource is small and expensive relatively.